### PR TITLE
New bounds

### PR DESCRIPTION
**Pull request by leclere**
* allow varying time bound constraint, working in sddp, deteriorated in sdp
* adapt examples and improve comments
* added subgradient recovering function

### CODE REVIEW

**Comment by leclere** [comment]
At the moment the Cut Pruning and integer are not working.

File: examples/stock-example.jl [hunk 1]
```diff
@@ -55,7 +55,7 @@ end
 
 # Define cost corresponding to each timestep:
 function cost_t(t, x, u, w)
-    return COSTS[t] * u[1]
+    return COSTS[t]* u[1] #
```
Comment (frapac) [inline examples/stock-example.jl:58]: Can we remove the trailing # ?

File: src/SDDPoptimize.jl [hunk 2]
```diff
@@ -579,7 +579,9 @@ function get_lower_bound(model::SPModel, param::SDDPparameters,
                             V::Vector{PolyhedralFunction})
     return get_bellman_value(model, param, 1, V[1], model.initialState)
 end
-
+function get_lower_bound(sddp::SDDPInterface)
```
Comment (frapac) [inline src/SDDPoptimize.jl:582]: Why build a function similar as `lowerbound` ?

File: src/SDDPoptimize.jl [hunk 3]
```diff
@@ -641,3 +643,47 @@ function add_cuts_to_model!(model::SPModel, t::Int64, problem::JuMP.Model, V::Po
...
+    maxvalue = -Inf
+    index = 0
+    for i in 1:Vt.numCuts
+        lambda = vec(Vt.lambdas[i, :])
+        if Vt.betas[i] + dot(lambda, x) >= maxvalue
+            maxvalue = Vt.betas[i] + dot(lambda, x)
+            index = i
+        end
+    end
+return Vt.lambdas[index, :]
```
Comment (frapac) [inline src/SDDPoptimize.jl:688]: Indentation

File: src/objects.jl [hunk 4]
```diff
@@ -96,13 +99,45 @@ end
...
+        return repmat(bounds,1,ns)
+    elseif ndims(bounds) == 2 && size(bounds) == (nx,ns)
+        return bounds
+     else
+        error(variable, " bounds dimension should be ", nx," or (",nx,",",ns,")")
+     end
+ end
+
+
+function iswithinbounds(x, bounds::Array)
```
Comment (frapac) [inline src/objects.jl:122]: This function can be optimized (break for loop if test became negative)

File: src/oneStepOneAleaProblem.jl [hunk 5]
```diff
@@ -53,14 +55,34 @@ function solve_one_step_one_alea(model,
...
+            @objective(m, Min, model.costFunctions(m, t, x, u, xi) + alpha)
+        end
+
+    elseif isa(model.costFunctions, Vector{Function})
+        cost = getindex(m, :cost)
+        for i in 1:length(model.costFunctions)
+            @constraint(m, cost >= model.costFunctions[i](t, x, u, xi))
+        end
+        @objective(m, Min, cost + alpha)
+     end
```
Comment (frapac) [inline src/oneStepOneAleaProblem.jl:84]: This is a major slow down in the algorithm: the fact that we need to redefine the objective each time we call `one_step_one_alea` in any configuration is too much an overhead IMO. 

My benchmark shows that the performance decreases in `new_bounds`, compared to `master`: the execution time increases from 8s to 17s ...

File: src/oneStepOneAleaProblem.jl [hunk 6]
```diff
@@ -118,8 +142,9 @@ function solve_dh(model, param, t, xt, m; verbosity::Int64=0)
     status = solve(m)
     solved = status == :Optimal
     if ~solved
-        println(m)
-        error("Foo")
+        warn("dh model not solved at time t=",t)
+        #println(m)
+        #error("Foo")
```
Comment (frapac) [inline src/oneStepOneAleaProblem.jl:147]: Can we remove the commented lines?

File: src/sdp.jl [hunk 7]
```diff
@@ -106,9 +106,9 @@ function build_sdpmodel_from_spmodel(model::SPModel)
     end
 
     if isa(model,LinearSPModel)
-        function cons_fun(t,x,u,w)
-            return true
-        end
+        
+        cons_fun(t,x,u,w) = true
```
Comment (frapac) [inline src/sdp.jl:110]: Trailing whitespace :)

**Review by frapac — CHANGES_REQUESTED** [review]
I am working on fixing the cut selection algorithm

File: src/sdp.jl (hunk 7 above)
Comment (trigaut) [reply src/sdp.jl:110]: I have to reconfigure sublime sorry ;)

**Comment by frapac** [comment]
I fix the problem with cuts pruning. 

Otherwise, I confirm that QP regularization works, but SMIP is broken ...

**Pull request closed without merging**